library/std/src/sys/sync/condvar/pthread.rs

```diff
@@ -1,196 +1,88 @@
-
-        cfg_if::cfg_if! {
-            if #[cfg(any(
-                target_os = "l4re",
-                target_os = "android",
-                target_os = "redox",
-                target_vendor = "apple",
-            ))] {
-                // `pthread_condattr_setclock` is unfortunately not supported on these platforms.
-            } else if #[cfg(any(target_os = "espidf", target_os = "horizon", target_os = "teeos"))] {
-                // NOTE: ESP-IDF's PTHREAD_COND_INITIALIZER support is not released yet
-                // So on that platform, init() should always be called
-                // Moreover, that platform does not have pthread_condattr_setclock support,
-                // hence that initialization should be skipped as well
-                //
-                // Similar story for the 3DS (horizon).
-                let r = unsafe { libc::pthread_cond_init(condvar.0.get(), crate::ptr::null()) };
-                assert_eq!(r, 0);
-            } else {
-                use crate::mem::MaybeUninit;
-                let mut attr = MaybeUninit::<libc::pthread_condattr_t>::uninit();
-                let r = unsafe { libc::pthread_condattr_init(attr.as_mut_ptr()) };
-                assert_eq!(r, 0);
-                let r = unsafe { libc::pthread_condattr_setclock(attr.as_mut_ptr(), libc::CLOCK_MONOTONIC) };
-                assert_eq!(r, 0);
-                let r = unsafe { libc::pthread_cond_init(condvar.0.get(), attr.as_ptr()) };
-                assert_eq!(r, 0);
-                let r = unsafe { libc::pthread_condattr_destroy(attr.as_mut_ptr()) };
-                assert_eq!(r, 0);
-            }
-        }
-
-        condvar
-    }
-}
-
-impl Drop for AllocatedCondvar {
-    #[inline]
-    fn drop(&mut self) {
-        let r = unsafe { libc::pthread_cond_destroy(self.0.get()) };
-        if cfg!(target_os = "dragonfly") {
-            // On DragonFly pthread_cond_destroy() returns EINVAL if called on
-            // a condvar that was just initialized with
-            // libc::PTHREAD_COND_INITIALIZER. Once it is used or
-            // pthread_cond_init() is called, this behavior no longer occurs.
-            debug_assert!(r == 0 || r == libc::EINVAL);
-        } else {
-            debug_assert_eq!(r, 0);
-        }
-    }
+    cvar: OnceBox<pal::Condvar>,
+    mutex: AtomicUsize,
 }
 
 impl Condvar {
     pub const fn new() -> Condvar {
-        Condvar { inner: OnceBox::new(), mutex: AtomicPtr::new(ptr::null_mut()) }
+        Condvar { cvar: OnceBox::new(), mutex: AtomicUsize::new(0) }
     }
 
-    fn get(&self) -> *mut libc::pthread_cond_t {
-        self.inner.get_or_init(AllocatedCondvar::new).0.get()
+    #[inline]
+    fn get(&self) -> Pin<&pal::Condvar> {
+        self.cvar.get_or_init(|| {
+            let mut cvar = Box::pin(pal::Condvar::new());
+            // SAFETY: we only call `init` once, namely here.
+            unsafe { cvar.as_mut().init() };
+            cvar
+        })
```

Comment:
So, as I understand it, the closure can potentially get called twice, as the sync can race. That means that `init` *has to be safe to call twice*, even if it means that it can be "safe, but only if you immediately destroy it if it's a copy". That means that this `// SAFETY` comment is somewhat off-base.

Reply:
The invariant is that `init` may only be called once per object; that isn't violated by this.

Reply:
Then perhaps I need a remedial lesson on synchronization here, because the comments in OnceBox are very sparing and I'm not sure how the sync works out:
- https://github.com/rust-lang/rust/blob/c322cd5c5ac13b5ef16a487c889f8c148d8e3841/library/std/src/sys/sync/once_box.rs#L45-L73

Reply:
I pried it apart.

I don't like "per object" as a constraint, because we have objects that manage objects, and we may create two objects to put inside one object, and then we have to get rid of one object, but we keep the other object, because we put it inside of the object...

... ... ...you get the picture.

What matters is that _the pointee_ only sees one init.